In [1]:
import numpy as np

In [2]:
import pandas as pd


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.0 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/traitlets/config/application.py", line 992, in launch_instance
    app.start()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel/kernelapp.py", line 711, in start
    self.io_

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.0 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/traitlets/config/application.py", line 992, in launch_instance
    app.start()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel/kernelapp.py", line 711, in start
    self.io_

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.0 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/traitlets/config/application.py", line 992, in launch_instance
    app.start()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel/kernelapp.py", line 711, in start
    self.io_

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.0 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/traitlets/config/application.py", line 992, in launch_instance
    app.start()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel/kernelapp.py", line 711, in start
    self.io_

AttributeError: _ARRAY_API not found

In [3]:
df = pd.read_csv("ab_test_synthetic_data.csv")

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   user_id    5000 non-null   int64 
 1   group      5000 non-null   object
 2   signed_up  5000 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 117.3+ KB


In [6]:
df['signed_up'].unique()

array([0, 1])

### STEP 1

**Business Hypothesis:**
We believe that shortening the sign-up form (shown to Group B) will reduce user friction and lead to more users signing up compared to the current longer form (Group A).

**Primary Metric:**
Sign-Up Rate = Number of users who signed up / Total users in the group

This is a binary metric (yes = 1, no = 0), so we’ll compare proportions using a z-test later on.

**Statistical Hypotheses:**
Null Hypothesis (H₀):
There is no difference in sign-up rates between the two groups.

Conversion_A = Conversion_B

Alternative Hypothesis (H₁):
There is a difference in sign-up rates between the two groups.

Conversion_A ≠ Conversion_B

This will be a two-tailed test, because we’re testing for any difference (not just improvement).



### STEP 2

**Power Analysis:**
Goal: Make sure the test can detect a real difference if it exists.

Power (1 - β): 80% (standard)
This means we accept a 20% chance of missing a real effect (Type II error).

Significance Level (α): 0.05 (5%)
We accept a 5% chance of false positive (Type I error).

**2. Minimum Detectable Effect (MDE)**
This is the smallest difference in sign-up rates that we care about detecting.

Let’s say the business cares about at least a 2% absolute improvement (from 12% to 14%)

So, MDE = 2%

**3. Sample Size (Pre-Defined)**
Based on our assumptions (α = 0.05, Power = 0.8, MDE = 2%), an online calculator gives us approximately:

~2,400 users per group

Total = 4,800 users (we generated 5,000 users in our synthetic dataset — this covers our requirement!)

**4. Test Duration:**
Assume:

Website gets 1,000 users/day
Estimated test duration = 5 days

(We don’t need to simulate time here — just important to keep in mind for real-world experiments.)

### STEP 3

Group A (Control):

Total users: 2,504

Sign-ups: 306

Conversion rate: 12.22%

Group B (Treatment):

Total users: 2,496

Sign-ups: 323

Conversion rate: 12.94%

There’s a slight improvement in Group B's conversion rate — about a 0.72% lift.



In [7]:
# Calculating the signup rate (conversion rate) for each group
conversion_A = df[df['group'] == 'A']['signed_up'].mean() * 100
conversion_B = df[df['group'] == 'B']['signed_up'].mean() * 100

print(f"Signup Rate for Group A: {conversion_A:.2f}%")
print(f"Signup Rate for Group B: {conversion_B:.2f}%")


Signup Rate for Group A: 12.22%
Signup Rate for Group B: 12.94%


### STEP 4: 

We’ve seen that:

Group A: 12.22% sign-up rate

Group B: 12.94% sign-up rate

That’s a small lift — but is it significant, or could it have happened just by luck?

To find out, we’ll use a z-test for proportions.

In [8]:
from statsmodels.stats.proportion import proportions_ztest

# Getting the counts of signups and total users
signups = [df[df['group'] == 'A']['signed_up'].sum(), df[df['group'] == 'B']['signed_up'].sum()]
totals = [df[df['group'] == 'A']['signed_up'].count(), df[df['group'] == 'B']['signed_up'].count()]

# Running a two-tailed z-test
z_stat, p_value = proportions_ztest(count=signups, nobs=totals, alternative='two-sided')

print(f"Z-statistic: {z_stat:.3f}")
print(f"P-value: {p_value:.4f}")


Z-statistic: -0.768
P-value: 0.4426


Out  p-value here (0.4426) is much greater than the standard threshold of 0.05

This means the difference between Group A and Group B's sign-up rates is not statistically significant

In other words: The 0.72% lift in Group B could very well be due to random chance

**Conclusion:**

We cannot reject the null hypothesis.
There isn’t enough statistical evidence to say that the shorter sign-up form (Group B) performs differently from the original (Group A).

### STEP 5: 

What Are We Asking Here?
Even though the difference in sign-up rates between Group A and B is not statistically significant, we now ask:

“Is the observed difference still meaningful enough to act on — from a business perspective?”

ince the test was not statistically significant, we cannot confidently say the observed improvement wasn’t just random.

Even though there’s a slight lift:

It’s not reliable enough to justify rollout

Best to run a follow-up test with a larger sample or refined variant


**Final Recommendation:**

Consider re-testing or improving the variant

